googleapi/main.js

```diff
@@ -0,0 +1,112 @@
+const electron = require('electron');
+const { BrowserWindow, ipcMain } = electron;
+
+const constants = require('./constants');
+
+/**
+ * Options to pass into new {@code BrowserWindow} instances if none are
+ * specified.
+ *
+ * @private
+ * @type {Object}
+ */
+const defaultBrowserWindowOptions = {
+    webPreferences: {
+        nodeIntegration: false
+    }
+};
+
+/**
+ * The configured {@code BrowserWindow} options to pass in to new BrowserWindow
+ * instances. This value can be overridden in {@code setupGoogleApiMain}.
+ *
+ * @private
+ * @type {Object}
+ */
+let browserWindowOptions = defaultBrowserWindowOptions;
+
+/**
+ * A reference to the opened Google authentication window. Kept so the window
+ * can be closed on call to {@code teardownGoogleApiMain}.
+ *
+ * @private
+ * @type {BrowserWindow|null}
+ */
+let googlePopup = null;
+
+/**
+ * Opens a new {@code BrowserWindow} instance for Google authentication and
+ * listens for redirects to a specified redirect URI.
+ *
+ * @param {Object} authRequestEvent - The event object passed from Electron.
+ * @param {Object} options - The additional arguments from the event.
+ * @param {string} options.authUrl - The URL to display for authentication.
+ * @param {string} options.redirectUri - The URL that is expected to be
+ * displayed after successful authentication.
+ * @private
+ * @returns {void}
+ */
+function onAuthRequested(authRequestEvent, options) {
+    if (googlePopup) {
+        googlePopup.focus();
+        return;
+    }
+
+    googlePopup = new BrowserWindow(browserWindowOptions);
+
+    googlePopup.loadURL(options.authUrl);
+
+    googlePopup.on('closed', () => {
+        // Fire this again in case the user closed the modal without finishing
+        // authentication.
+        authRequestEvent.sender.send(constants.AUTH_FINISHED);
+
+        googlePopup = null;
+    });
+
+    googlePopup.webContents.on('did-navigate', (navEvent, url) => {
+        if (url.indexOf(options.redirectUri) === 0) {
+            authRequestEvent.sender.send(constants.AUTH_FINISHED, url);
+            googlePopup.close();
+        }
+    });
+}
+
+/**
+ * Sets listeners on the main Electron process to handle requests to open a
+ * {@code BrowserWindow} instance for Google authentication.
+ *
+ * @param {Object} options - Additional configuration for setting up the
+ * listeners.
+ * @param {Object} browserWindowOptions - The options to pass into the
+ * constructor for {@code BrowserWindow}.
+ * @returns {void}
+ */
+function setupGoogleApiMain(options = {}) {
+    teardownGoogleApiMain();
+
+    browserWindowOptions
```

Comment:
Actually why do we need to expose the BrowserWindow options? Do you have an use case in mind where the consumer of the library will need to change anything there?

Reply:
I don't know what the consumer needs. For example, the consumer might have security restrictions on what options needs to be set on new windows. So to reduce the need to make changes in utils I just opened up options.

Reply:
I'm with mixed feelings about this but this make sense.

Reply:
I'm open to changes here. Request them publicly or privately. I can make some assumptions of what will be used by the consumer I'm building this out for specifically.

Reply:
I don't have better idea currently. Let's leave like it is!